gorm/transaction.go

```diff
@@ -57,15 +90,50 @@ func (t *Transaction) AddAfterCommitHook(hooks ...func(context.Context)) {
 	t.afterCommitHook = append(t.afterCommitHook, hooks...)
 }
 
-// BeginFromContext will extract transaction wrapper from context and start new transaction.
+// getReadOnlyDBInstance returns the read only database instance stored in the ctx
+func getReadOnlyDBInstance(ctx context.Context) (*gorm.DB, error) {
+	txn, ok := FromContext(ctx)
+	if !ok {
+		return nil, ErrCtxTxnMissing
+	}
+	dbRO, ok := ctx.Value(roDBKey).(*gorm.DB)
+	if !ok {
+		logger.Warnf("BeginFromContext: requested: read-only DB, returns: read-write DB, reason: read-only DB not available")
+		if txn.parent == nil {
+			return nil, ErrCtxTxnNoDB
+		}
+		db := txn.beginWithContext(ctx)
+		if db.Error != nil {
+			return nil, db.Error
+		}
+		return db, nil
+	}
+	txn.readOnly = true
+	return dbRO, nil
+}
+
+// BeginFromContext will return read only db instance if readOnlyReplica flag is set otherwise it will extract transaction wrapper from context and start new transaction
+// If readOnlyReplica flag is set and a txn with read-write db is already in use then it will return a txn from ctx rather than providing a read-only db instance.
 // As result new instance of `*gorm.DB` will be returned.
 // Error will be returned in case either transaction or db connection info is missing in context.
 // Gorm specific error can be checked by `*gorm.DB.Error`.
-func BeginFromContext(ctx context.Context) (*gorm.DB, error) {
+func BeginFromContext(ctx context.Context, options ...DatabaseOption) (*gorm.DB, error) {
 	txn, ok := FromContext(ctx)
 	if !ok {
 		return nil, ErrCtxTxnMissing
 	}
+	opts := toDatabaseOptions(options...)
+	if opts.readOnlyReplica == true {
+		if txn.current == nil {
+			return getReadOnlyDBInstance(ctx)
+		} else {
+			logger.Warnf("BeginFromContext: requested: read-only DB, returns: read-write DB, reason: read-write DB txn in use")
+			return txn.current, nil
+		}
+	} else if txn.readOnly == true {
+		logger.Warnf("BeginFromContext: requested: read-write DB, returns: read-only DB, reason: txn set to read only")
+		return getReadOnlyDBInstance(ctx)
+	}
```

Comment:
Returning a R/W transaction instead of R-O transaction is ok since all the read-only operations will be executed by the R/W transaction. Doing the inverse doesn't work. This should return an error.